### PR TITLE
[Instrumentation.AspNet] pass http.request.method to sampler

### PR DESCRIPTION
**Pull request by roycald245**
Fixes [#57287](https://github.com/dotnet/aspnetcore/issues/57287)

## Changes
Added http.request.method to tags given to sampler

## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/CONTRIBUTING.md) guidelines followed (license requirements, nullable enabled, static analysis, etc.)
* [x] Unit tests added/updated
* [x] Appropriate `CHANGELOG.md` files updated for non-trivial changes
* [x] Changes in public API reviewed (if applicable)

### CODE REVIEW

File: src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs
Comment (Kielek) [inline src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs:72]: I doubt that that the `path` should be passed as a value parameter here.

Comment (roycald245) [reply src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs]: You are correct... fixed in [f62e71c](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/2023/commits/f62e71cb50a74305f654b363c3571a34cfd2afdc)

File: src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/CHANGELOG.md
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/CHANGELOG.md:12]: please merge this into the previous entry, so we don't need to duplicate most of the contents

Comment (roycald245) [reply src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/CHANGELOG.md]: fixed in [4c3c59a](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/2023/commits/4c3c59a957154b687e29ccb72c76ec55fcfd504a)

File: src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/CHANGELOG.md
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/CHANGELOG.md:5]:
```suggestion
* `TelemetryHttpModule` will now pass the `http.request.method` and the `url.path` tags (set to
```

File: src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/CHANGELOG.md
Comment (Kielek) [inline src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/CHANGELOG.md:12]:
```suggestion
  ([#1871](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1871),
   [#2023](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/2023))
```

File: src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs
Comment (vishweshbankwar) [inline src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs:68]: I think if we need to set the normalized value as we do [here](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/7fcae4903133fb8d6e249963bceaa81689c5e326/src/OpenTelemetry.Instrumentation.AspNet/Implementation/HttpInListener.cs#L77)

If the method is normalized to `_OTHER` then should we also include [http.request.method_original](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/attributes-registry/http.md), that part is not very clear in [spec](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/http/http-spans.md#http-server:~:text=implementations%20are%20possible.-,The%20following%20attributes%20can%20be%20important%20for%20making%20sampling%20decisions%20and%20SHOULD%20be%20provided%20at%20span%20creation%20time%20(if%20provided%20at%20all)%3A,-client.address)

Comment (roycald245) [reply src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs]: fixed [34e7fa3](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/2023/commits/34e7fa38970f0b86c9e184540895b46e59534e8a)

Comment (roycald245) [reply src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs]: I think setting "HTTP" like the example you sent is the way to go. Keeping the convention.

File: src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs
Comment (Kielek) [inline src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs:23]: Why false? This package in other places is using true as parameter.

```suggestion
    internal static readonly RequestDataHelper RequestDataHelper = new(configureByHttpKnownMethodsEnvironmentalVariable: true);
```

Comment (roycald245) [reply src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs:23]: I took TelemetryHelper within AspNetCore instrumentation for example, in which it is false.
But with further reading of the parameter it seems like passing true is the correct way to go.

Comment (roycald245) [reply src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs:23]: It is also false in OpenTelemetry.Instrumentation.Http

Comment (Kielek) [reply src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs]: You are right, there is no possibility to configure it in AspNetCore. As I remember it is not supported by .NET8+ with native instrumentations.

Comment (roycald245) [reply src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs]: Okay, since I already changed it to true I will revert it :)

Comment (Kielek) [reply src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs:23]: Please revert your revert. It is not supported on .NET8 for AspNetCore.
AspNet is something different, ans the parameter should be true!

**Comment by birojnayak** [comment]
@cijothomas @Kielek this would be really useful for us.... can we have it available sooner

File: src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs
Comment (ysolomchenko) [inline src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs:67]: Why not replace `KeyValuePair<string, object?>[]?` with `List<KeyValuePair<string, object?>>?` for dynamic tag management, which eliminates the need for manual index handling in arrays, making the code more flexible and cleaner?

Comment (Kielek) [reply src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs:67]: @ysolomchenko, I would say that it is due to performance. If you think that it is worth to change without loosing it, please create PR. Some benchmarks will be needed.

**Comment by Kielek** [comment]
@roycald245, one test is failing.

File: src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs
Comment (ysolomchenko) [reply src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs:67]: @Kielek, Yes, the speed will indeed drop by about five times (39 ns vs. 7 ns). However, since we're talking about nanoseconds, I'm not sure how critical that is. It might not be noticeable in real-world scenarios. If this difference is truly important, then we should consider abandoning this change.

Comment (Kielek) [reply src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs:67]: It is on the hot-path for each asp.net request. If it will be 5-6 times slower, we should avoid switching to the list version.

**Comment by roycald245** [comment]
I had some distractions recently. Will fix as soon as I'm able

**Comment by Kielek** [comment]
@roycald245, I have made some direct changes to
- handle http method attributes as in RequestDataHelper.SetHttpMethodTag
- fix cashing tags storage
- extend tests

Please double check if you are fine with this solution.

@cijothomas, could you please review?

**Review by ysolomchenko — APPROVED** [review]
LGTM

**Comment by roycald245** [comment]
Need codeowner approval

File: src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs
Comment (CodeBlanch) [inline src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs:80]: * We can probably remove this also being done in AspNet because sampler tags are automatically added to the span IIRC: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/32c4e943405acf811120b9723aabd3740f63b8ae/src/OpenTelemetry.Instrumentation.AspNet/Implementation/HttpInListener.cs#L77

* Consider refactoring this into `RequestDataHelper` something like this...

```csharp
RequestDataHelper.SetHttpMethodTag(tags, originalHttpMethod: context.Request.HttpMethod);

class RequestDataHelper
{
    public void SetHttpMethodTag(List<KeyValuePair<string, object?>> tags, string originalHttpMethod)
    {
        var normalizedHttpMethod = this.GetNormalizedHttpMethod(originalHttpMethod);
        tags.Add(new(SemanticConventions.AttributeHttpRequestMethod, normalizedHttpMethod));

        if (originalHttpMethod != normalizedHttpMethod)
        {
            tags.Add(new(SemanticConventions.AttributeHttpRequestMethodOriginal, originalHttpMethod));
        }
    }
}
```

Reason being it seems like all of this logic is already encapsulated inside `RequestDataHelper`.

File: src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs
Comment (CodeBlanch) [inline src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs:69]: Is it possible for `context` to be `null` here?

```suggestion
        string? path = context.Request.Unvalidated?.Path;
        string originalHttpMethod = context.Request.HttpMethod;
```

**Comment by Kielek** [comment]
@roycald245, do you have some time to address @CodeBlanch comments?

**Pull request closed without merging**
